pkgs/development/interpreters/octave/mk-octave-derivation.nix

```diff
@@ -0,0 +1,69 @@
+# Generic builder for GNU Octave libraries.
+# This is a file that contains nested functions. The first, outer, function
+# is the library- and package-wide details, such as the nixpkgs library, any
+# additional configuration provided, and the namePrefix to use (based on the
+# pname and version of Octave), the octave package, etc.
+
+{ pkgs
+, lib
+, stdenv
+, config
+, namePrefix
+, octave
+}:
+
+# The inner function contains information required to build the individual
+# libraries.
+{ fullLibName ? "${attrs.pname}-${attrs.version}"
+
+, src
+
+# Build-time dependencies for the package
+, nativeBuildInputs ? []
+
+# Run-time dependencies for the package
+, buildInputs ? []
+
+# Propagate build dependencies so in case we have A -> B -> C,
+# C can import package A propagated by B
+, propagatedBuildInputs ? []
+
+, meta ? {}
+
+, passthru ? {}
+
+, ... } @ attrs:
+
+let
+  self = stdenv.mkDerivation {
+    packageName = "${fullLibName}";
+    name = "${namePrefix}-${fullLibName}";
+    src = src;
+
+    OCTAVE_HISTFILE = "/dev/null";
+
+    dontUnpack = true;
+
+    nativeBuildInputs = [
+      octave
```

Comment:
I think this might warrant a comment? I wonder how python's packages handle cross compiling, where this choice might not be what we want..

Reply:
I am not quite sure what you are referring to with this snippet here. What warrants a comment? What about python?

Reply:
> What warrants a comment?

In general, `nativeBuildInputs` is different then `buildInputs`. I don't know how to best explain this, but this makes a difference when we consider cross compiling. Currently, you don't differentiate between the octave derivation that should be used on the "host" platform and the "build" platform. This is not that significant since our octave is anyway only for x86_64 I think, but we might want to compose things in a way that if nixpkgs will gain support for compiling octave on other platforms, the `octave.pkgs` will be able to be cross compiled. IIRC octave has support for cross compiling upstream.

I'm not suggesting with this comment how things can be better, I'm just mentioning this here, because someone more experienced than me with cross compiling, will have a good say about this. So let's keep this thread open :).

Reply:
AH! That's what you meant! I wasn't quite sure what you were referring to before.

When writing the mk-octave-derivation.nix file, I was mirroring python's [mk-python-derivation.nix](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/interpreters/python/mk-python-derivation.nix). So I wasn't quite sure what I was doing (and I still don't).

But, octave can already be built for other architectures. It looks that Debian has an aarch64 version of octave (see [here](https://packages.debian.org/sid/arm64/octave/filelist)).

Reply:
So instead of having `nativeBuildInputs = [ octave ]`, have `buildInputs = [ octave ]`, @FRidh?

Reply:
No, keep it. Eventually for cross it would be `nativeBuildInputs = [ buildPackages.octave ]; `.